vividus-engine/src/main/java/org/vividus/annotation/ReplacementMonitor.java

```diff
@@ -0,0 +1,112 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.vividus.annotation;
+
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.Formatter;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.jbehave.core.configuration.Keywords;
+import org.jbehave.core.model.StepPattern;
+import org.jbehave.core.steps.NullStepMonitor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.vividus.context.RunTestContext;
+
+public class ReplacementMonitor extends NullStepMonitor
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(ReplacementMonitor.class);
+
+    private final RunTestContext runTestContext;
+
+    private final Map<String, String> stepPatternsCache = new ConcurrentHashMap<>();
+
+    public ReplacementMonitor(RunTestContext runTestContext)
+    {
+        this.runTestContext = runTestContext;
+    }
+
+    @Override
+    public void stepMatchesPattern(String step, boolean matches, StepPattern pattern, Method method,
+            Object stepsInstance)
+    {
+        if (matches)
+        {
+            String key = pattern.type() + " " + pattern.annotated();
+            stepPatternsCache.putIfAbsent(key, pattern.resolved());
+        }
+    }
+
+    @Override
+    public void beforePerforming(String step, boolean dryRun, Method method)
+    {
+        getAnnotation(method).ifPresent(r -> {
+            String rawDeprecatedStep = runTestContext.getRunningStory().getRunningSteps().getFirst();
+            Matcher stepMatcher = getStepMatcher(rawDeprecatedStep);
+            String[] stepParams = getParametersFromStep(stepMatcher);
+
+            try (Formatter newStep = new Formatter())
+            {
+                newStep.format(r.replacementFormatPattern(), stepParams);
+
+                LOGGER.atInfo()
+                        .addArgument(rawDeprecatedStep)
+                        .addArgument(r::versionToRemoveStep)
+                        .addArgument(newStep::toString)
+                        .log("The step: \"{}\" is deprecated and will be removed in VIVIDUS {}. Use step: \"{}\"");
+            }
+        });
+    }
+
+    private Optional<Replacement> getAnnotation(Method method)
+    {
+        return Optional.ofNullable(method).map(m -> method.getAnnotation(Replacement.class));
+    }
+
+    private Matcher getStepMatcher(String rawDeprecatedStep)
+    {
+        List<Matcher> matchers = new ArrayList<>();
+        for (String stepPattern : stepPatternsCache.values())
+        {
+            Keywords keywords = new Keywords();
+            Pattern pattern = Pattern.compile(stepPattern, Pattern.DOTALL);
```

Comment:
we always re compile stepPattern, does it make sense to cache it somehow?

Reply:
I think exactly in this case we can really optimize the logic without complicating the code
Since we are looking for all possible patterns, it makes sense to compile them with the cache initialization